AUTHORING_GUIDE.md

```diff
@@ -281,216 +191,203 @@ Args:
+Sample code should use standard Python exception handling techniques as
+covered in the [Google Python Style
+Guide](http://google.github.io/styleguide/pyguide.html#24-exceptions).
 
-Command-line arguments should generally have a 1-to-1 match to function
-arguments. For example:
+# Testing Guidelines
 
-```python
-parser.add_argument('source_file_name')
-parser.add_argument('destination_blob_name')
-```
+Samples should include tests to verify that the sample runs correctly and
+generates the intended output. Follow these guidelines while writing your
+tests:
+
+* Use [pytest](https://docs.pytest.org/en/latest/)-style tests and plain
+asserts. Don't use `unittest`-style tests or `assertX` methods.
+* Whenever possible, tests should avoid taking a dependency on the precise
+output of APIs. For example, if a test is intended to verify a JSON payload
+returned from an endpoint, it should only check for the existence of the
+expected keys and values, and the test should continue to work correctly
+if other values are added to the response in a future version of the API.
+* Samples that use App Engine Standard should use the [App Engine
+testbed](https://cloud.google.com/appengine/docs/standard/python/refdocs/google.appengine.ext.testbed)
+for system testing, as shown in [this
+example](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/appengine/standard/localtesting/datastore_test.py).
+* All tests should be independent of one another and order-independent.
 
-Again, descriptive names prevent you from having to exhaustively describe
-every parameter.
+## Arrange, Act, Assert
 
-Some samples demonstrate multiple functions. You should use *subparsers* to
-handle this, for example:
+Tests for samples should follow the “Arrange, Act, Assert” structure for
+unit tests:
 
-```python
-if __name__ == '__main__':
-    parser = argparse.ArgumentParser(
-        description=__doc__,
-        formatter_class=argparse.RawDescriptionHelpFormatter)
-    parser.add_argument('bucket_name', help='Your cloud storage bucket.')
+* _Arrange_ — create and configure the components required for the test.
+Avoid nesting; prioritize readability and simplicity over efficiency.
+* _Act_ — execute the code to be tested, such as sending a request and
+receiving a response.
+* _Assert_ — verify that the test results match what is expected.
 
-    subparsers = parser.add_subparsers(dest='command')
+## External Resources
 
-    subparsers.add_parser('list', help=list_blobs.__doc__)
+Whenever possible, tests should run against the live production version of
+cloud APIs and resources. This will assure that any breaking changes in those
+resources are identified by the tests.
 
-    upload_parser = subparsers.add_parser('upload', help=upload_blob.__doc__)
-    upload_parser.add_argument('source_file_name')
-    upload_parser.add_argument('destination_blob_name')
+Creating mocks for external resources is optional, and not recommended in
```

Comment:
I'd like this wording to be stronger - Mocks should be strongly discouraged, as they don't verify the validity of the sample against the API.

Reply:
Added note about use of mocks strongly discouraged, tests should verify sample against the actual APIs. Addressed in commit [751c8e2](https://github.com/GoogleCloudPlatform/python-docs-samples/pull/2728/commits/751c8e27c591a489d2cb47477b0a971d41fcf36a)